src/pkg/plug/scanner/api/registration.go

```diff
@@ -0,0 +1,170 @@
+}
+
+// CreateRegistration ...
+func (bc *basicController) CreateRegistration(registration *dscanner.Registration) (string, error) {
+	// TODO: Get metadata from the adapter service first
+	return bc.manager.Create(registration)
+}
+
+// GetRegistration ...
+func (bc *basicController) GetRegistration(registrationUUID string) (*dscanner.Registration, error) {
+	return bc.manager.Get(registrationUUID)
+}
+
+// RegistrationExists ...
+func (bc *basicController) RegistrationExists(registrationUUID string) bool {
+	registration, err := bc.manager.Get(registrationUUID)
+
+	return !(err == nil && registration == nil)
+}
+
+// UpdateRegistration ...
+func (bc *basicController) UpdateRegistration(registration *dscanner.Registration) error {
+	return bc.manager.Update(registration)
+}
+
+// SetDefaultRegistration ...
+func (bc *basicController) DeleteRegistration(registrationUUID string) (*dscanner.Registration, error) {
+	registration, err := bc.manager.Get(registrationUUID)
+	if registration == nil && err == nil {
+		// Not found
+		return nil, nil
+	}
+
+	if err := bc.manager.Delete(registrationUUID); err != nil {
+		return nil, errors.Wrap(err, "delete registration")
+	}
+
+	return registration, nil
+}
+
+// SetDefaultRegistration ...
+func (bc *basicController) SetDefaultRegistration(registrationUUID string) error {
+	return bc.manager.SetAsDefault(registrationUUID)
+}
+
+// SetRegistrationByProject ...
+func (bc *basicController) SetRegistrationByProject(projectID int64, registrationID string) error {
+	if projectID == 0 {
+		return errors.New("invalid project ID")
+	}
+
+	if len(registrationID) == 0 {
+		return errors.New("missing scanner UUID")
+	}
+
+	// Only keep the UUID in the metadata of the given project
+	// Scanner metadata existing?
+	m, err := bc.proMetaMgr.Get(projectID, proScannerMetaKey)
+	if err != nil {
+		return errors.Wrap(err, "set project scanner")
+	}
+
+	// Update if exists
+	if len(m) > 0 {
+		// Compare and set new
+		if registrationID != m[proScannerMetaKey] {
+			m[proScannerMetaKey] = registrationID
+			if err := bc.proMetaMgr.Update(projectID, m); err != nil {
+				return errors.Wrap(err, "set project scanner")
+			}
+		}
+	} else {
+		meta := make(map[string]string, 1)
+		meta[proScannerMetaKey] = registrationID
+		if err := bc.proMetaMgr.Add(projectID, meta); err != nil {
+			return errors.Wrap(err, "set project scanner")
+		}
+	}
+
+	return nil
+}
+
+// GetRegistrationByProject ...
+func (bc *basicController) GetRegistrationByProject(projectID int64) (*dscanner.Registration, error) {
+	if projectID == 0 {
+		return nil, errors.New("invalid project ID")
+	}
+
+	// First, get it from the project metadata
+	m, err := bc.proMetaMgr.Get(projectID, proScannerMetaKey)
+	if err != nil {
+		return nil, errors.Wrap(err, "get project scanner")
+	}
+
+	if len(m) > 0 {
+		if registrationID, ok := m[proScannerMetaKey]; ok && len(registrationID) > 0 {
+			registration, err := bc.manager.Get(registrationID)
```

Comment:
When registration of registrationID deleted in the system, maybe return the default one?

Reply:
Not get your point.

Reply:
eg, for project `library` set the scanner `clair` before, then system administrator delete the scanner, this function will return `nil` scanner.

Reply:
Actually that's the logic of this method. Seems there still are some gaps. Will enhance the code to be more strong.

Reply:
```go
if len(m) > 0 {
		if registrationID, ok := m[proScannerMetaKey]; ok && len(registrationID) > 0 {
			registration, err := bc.manager.Get(registrationID)
			if err != nil {
				return nil, errors.Wrap(err, "get project scanner")
			}

			if registration == nil {
				// Not found
				// Might be deleted by the admin, the project scanner ID reference should be cleared
				if err := bc.proMetaMgr.Delete(projectID, proScannerMetaKey); err != nil {
					return nil, errors.Wrap(err, "get project scanner")
				}
			} else {
				return registration, nil
			}
		}
	}```